geo/src/algorithm/sweep/line_or_point.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    #[inline]
+    pub fn end_points(&self) -> (SweepPoint<T>, SweepPoint<T>) {
+        (self.left, self.right)
+    }
+}
+
+/// Equality based on ordering defined for segments as per algorithm.
+impl<T: GeoNum> PartialEq for LineOrPoint<T> {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.partial_cmp(other) == Some(Ordering::Equal)
+    }
+}
+
+/// Ordering defined for segments as per algorithm.
+///
+/// Requires the following conditions:
+///
+/// 1. If comparing two lines, the both left ends must be strictly
+/// smaller than both right ends.
+///
+/// 2. A point is treated as a infinitesimal small vertical segment
+/// centered at its coordinates.
+impl<T: GeoNum> PartialOrd for LineOrPoint<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match (self.is_line(), other.is_line()) {
+            (false, false) => {
+                if self.left == other.left {
+                    Some(Ordering::Equal)
+                } else {
+                    // Unequal points do not satisfy pre-condition and
+                    // can't be ordered.
+                    None
+                }
+            }
+            (false, true) => other.partial_cmp(self).map(Ordering::reverse),
+            (true, false) => {
+                let (p, q) = self.end_points();
+                let r = other.left;
+                if r > q || p > r {
+                    return None;
+                }
+                Some(
+                    T::Ker::orient2d(*p, *q, *r)
+                        .as_ordering()
+                        .then(Ordering::Greater),
+                )
+            }
+            (true, true) => {
+                let (p1, q1) = self.end_points();
+                let (p2, q2) = other.end_points();
+                if p1 > p2 {
+                    return other.partial_cmp(self).map(Ordering::reverse);
+                }
+                if p1 >= q2 || p2 >= q2 {
+                    return None;
+                }
+
+                // Assertion: p1 <= p2
+                // Assertion: pi < q_j
+                Some(
+                    T::Ker::orient2d(*p1, *q1, *p2)
+                        .as_ordering()
+                        .then_with(|| T::Ker::orient2d(*p1, *q1, *q2).as_ordering()),
+                )
+            }
+        }
+    }
+}
+
+impl<T: GeoFloat> LineOrPoint<T> {
+    /// Intersect a line with self and return a point, a overlapping segment or `None`.
+    ///
+    /// The `other` argument must be a line variant (panics otherwise).
+    pub fn intersect_line(&self, other: &Self) -> Option<Self> {
+        debug_assert!(other.is_line(), "tried to intersect with a point variant!");
+
+        let line = other.line();
+        if !self.is_line() {
+            let p = self.left;
+            if <T as HasKernel>::Ker::orient2d(line.start, *p, line.end) == Orientation::Collinear {
```

Comment:
The conversions should get optimized to no-ops right?

Reply:
I'm not sure which conversions you're asking about.

To clarify, my take-it-or-leave-it suggestion was to replace the above with something like:

```rust
let line = other.line();
if !self.is_line() {
    let p = self.left;
    if line.intersects(&*p) {
        Some(*self)
    }  else {
        None
    }
} else {
    line_intersection(self.line(), line).map(|l| match l {
        LineIntersection::SinglePoint { intersection, .. } => intersection.into(),
        LineIntersection::Collinear { intersection } => intersection.into(),
    })
}
```

Which doesn't introduce any new conversions - I just liked it because it's more succinct.

Reply:
Makes sense; done!